UDP send coalescing + 'nodelay' message ID support

jkflying opened the pull request:
This is a packet coalescing implementation for UDP endpoints using time and size transmit thresholds. Before, especially when communicating over ethernet or wifi, each (on average) 40B Mavlink packet could end up in its own IP packet, amplifying traffic significantly. This can be worse on half-duplex links, where a single packet in one direction may prevent the frame from being used for critical communications in the other direction..

Delay, coalescing, nodelay mechanism is unit tested, and the file parsing for all 3 fields has been manually tested.

General design is:

    On receipt of data to send, push it into the TX buffer
    
    If the buffer has more data than Y, 
     or the timeout is set to 0,
     or if the message was in the NoDelay list,
        write immediately
    Else if there is no send event scheduled, schedule one to happen in X

Theoretically, setting the send size threshold to 0 should give exactly the same performance as before, minus a couple of comparisons and an extra memcpy per packet.

This feature can be accessed by adding the following parameters to the config file under a UDP endpoint (example values populated):

```
#1200 bytes is a bit less than 1500 for ethernet
CoalesceBytes = 1200

#60ms probably isn't noticeable for telemetry on the groundstation
CoalesceMs = 60

#these are the mavlink IDs for messages COMMAND_INT, COMMAND_LONG and COMMAND_ACK
CoalesceNoDelay = 75,76,77
```

Commit:
Remove trailing spaces

Commit:
override destructors instead of virtual

jkflying commented:
@nicovanduijn This should be good to go for higher delay values, but still getting low latency on commands.

caleridas commented on src/mavlink-router/endpoint.cpp:
```diff
@@ -761,22 +780,45 @@ ssize_t UdpEndpoint::_read_msg(uint8_t *buf, size_t len)
+    memcpy(&tx_buf.data[tx_buf.len], pbuf->data, pbuf->len);
+    tx_buf.len += pbuf->len;
+
+    if (!_write_scheduled) {
```
maybe move that check into _schedule_write, encapsulates a bit more

caleridas commented on src/mavlink-router/mainloop_test.cpp:
```diff
@@ -37,7 +37,7 @@ class MainLoopTest : public ::testing::Test {
     static std::pair<int, sockaddr_in> make_scratch_udp_socket()
     {
-        int fd = ::socket(AF_INET, SOCK_DGRAM | SOCK_CLOEXEC, 0);
+        int fd = ::socket(AF_INET, SOCK_DGRAM | SOCK_CLOEXEC | SOCK_NONBLOCK, 0);
```
would rather suggest explicitly using MSG_DONTWAIT in first recv below -- this way you get well-defined behavior in the second receive (which *should* block until delivered... just in case network stack may in some corner cases not deliver instantly -- I think this is well-defined for uds only, for udp less certain now)

jkflying replied:
It's also more explicit with DONTWAIT, good idea, I'll change it.

caleridas commented on src/mavlink-router/endpoint.cpp:
```diff
@@ -792,10 +798,15 @@ int UdpEndpoint::write_msg(const struct buffer *pbuf)
     tx_buf.len += pbuf->len;
 
-    if (!_write_scheduled) {
+    int ret = pbuf->len;
```
suggest to move this into "schedule_write", i.e. leave decision of whether to send immediately also to it (so schedule can mean "instant" or "later")

jkflying replied:
This gets tricky with trying to return sane values from the `write_msg` in case the caller is expecting values that would align with nonbuffered writes.

caleridas commented on src/mavlink-router/mainloop.h:
```diff
@@ -161,6 +161,8 @@ struct endpoint_config {
             int retry_timeout;
             bool eavesdropping;
+            int coalesce_ms;
+            int coalesce_bytes;
```
I think these two fields warrant description here -- config data structure would be a good place.

caleridas commented on src/mavlink-router/endpoint.cpp:
```diff
@@ -350,6 +350,22 @@ bool Endpoint::accept_msg(int target_sysid, int target_compid, uint8_t src_sysid
+void Endpoint::postprocess_msg(int target_sysid, int target_compid, uint8_t src_sysid,
+                               uint8_t src_compid, uint32_t msg_id)
+{
+    // avoid unused variable messages
```
commenting out formal parameter names would also accomplish this -- question of preferences probably

caleridas commented on src/mavlink-router/endpoint.cpp:
```diff
@@ -706,7 +722,7 @@ UdpEndpoint::UdpEndpoint()
         _max_timeout_ms, [this](void*)
             {
-                _force_write();
+                this->flush_pending_msgs();
```
unnecessary this->

caleridas commented on src/mavlink-router/mainloop.cpp:
```diff
@@ -494,11 +496,23 @@ bool Mainloop::add_endpoints(Mainloop &mainloop, struct options *opt)
+
+            if (conf->coalesce_nodelay) {
+                char *local_nodelay = strdup(conf->coalesce_nodelay);
+                char *token = strtok(local_nodelay, ",");
```
would advise to avoid C string handling, use std::string

jkflying replied:
Agreed in general... I just don't want to change too much with this PR

caleridas commented on src/mavlink-router/mainloop_test.cpp:
```diff
@@ -29,6 +29,9 @@ class MainLoopTest : public ::testing::Test {
             cfg.coalesce_ms = 1;
         }
         cfg.filter = nullptr;
+        static char nodelay[] = "75,76,77";
```
that is might to spark surprises -- since strtok is destructive, and it is called on this, this config string will be modified after first unit test and then run subsequent tests with different config

see suggestion in other comment

jkflying replied:
Yes, I've been bitten by that one before, so I made local copies when I did the strtok.

caleridas commented on src/mavlink-router/mainloop_test.cpp:
```diff
@@ -93,6 +93,15 @@ TEST_F(MainLoopTest, direct_udp_endpoint_send)
     char recvbuf[1024];
     ssize_t count = ::recv(sock, recvbuf, 1024, 0);
 
+    ASSERT_EQ(-1, count) << "UDP packet coalescing shouldn't have allowed a send yet";
```
would suggest preferred EXPECT -- ASSERT is primarily helpful for conditions that lead to memory corruption with very hard to understand follow-up errors

jkflying replied:
Yep, I think these crept in in the beginning when I was trying to avoid hanging the test. I'll switch back to EXPECT.

caleridas reviewed (approved):
looks good and tested -- minor comments inline

maybe it is possible to structure commits in a way that you avoid "changing code again" in multiple commits, perhaps something likethis:

- generic fixes (whitespace, ...)
- impl. with test, with implicit config defaulting to off
- config bits to expose functionality

Commit:
Add the capability of coalescing on UDP endpoints

In order to group small packets together, this can be done with both a timeout and
a size limit. Additionally, a list of 'high priority' message IDs can be given which
will cause the UDP endpoint to send immediately.

Commit:
Add file parsing for UDP endpoint packet coalescing.

This allows the configuration file to specify coalescing delay,
maximum coalescing packet size, and message IDs which have no delay
